File: src/runtime/context.js
import fs from "fs/promises";
import { createIgaHelpers } from "./iga.js";
import { createParameterReader } from "./params.js";

export async function loadRuntimeContext({ env = process.env, readFile = fs.readFile } = {}) {
  const contextFile = env.IGA_SCHEDULER_CONTEXT_FILE;
  if (!contextFile || typeof contextFile !== "string") throw runtimeContextError("RUNTIME_CONTEXT_FILE_REQUIRED", "IGA_SCHEDULER_CONTEXT_FILE is required");
  let raw;
  try { raw = await readFile(contextFile, "utf8"); } catch (error) { throw runtimeContextError("RUNTIME_CONTEXT_FILE_READ_FAILED", `failed to read runtime context file: ${error.message}`, { cause: error }); }
  let context;
  try { context = JSON.parse(raw); } catch (error) { throw runtimeContextError("RUNTIME_CONTEXT_JSON_INVALID", `runtime context JSON is invalid: ${error.message}`, { cause: error }); }
  if (!context || typeof context !== "object" || Array.isArray(context)) throw runtimeContextError("RUNTIME_CONTEXT_INVALID", "runtime context must be a JSON object");
  return context;
}

export async function createRuntimeContext(options = {}) {
  const context = await loadRuntimeContext(options);
  const params = context.params || {};
  return { raw: context, runId: context.runId, definition: context.definition, instance: context.instance, scheduledFireTime: context.scheduledFireTime, attempt: context.attempt, params, param: createParameterReader(params), iga: createIgaHelpers({ client: options.igaClient, bridge: options.igaBridge || createContextBridge(context.igaBridge) }) };
}

function createContextBridge(config) {
  if (!config) return undefined;
  return { invoke: async () => { const error = new Error("runtime IGA bridge transport is not configured in this SDK build"); error.code = "RUNTIME_IGA_BRIDGE_TRANSPORT_UNAVAILABLE"; throw error; } };
}

function runtimeContextError(code, message, { cause } = {}) {
  const error = new Error(message);
  error.code = code;
  if (cause) error.cause = cause;
  return error;
}

File: sdk/python/iga_scheduler/context.py
import json
import os

from .iga_client import resolve_iga_client


class ParameterReader:
    def __init__(self, params):
        self._params = params or {}

    def required_string(self, name):
        v = self._params.get(name)
        if not v or not isinstance(v, str):
            raise ValueError(f"missing required string parameter: {name}")
        return v

    def required_string_array(self, name):
        v = self._params.get(name)
        if not isinstance(v, list) or not v:
            raise ValueError(f"missing required string array parameter: {name}")
        return v


def create_context():
    context_file = os.environ.get("IGA_SCHEDULER_CONTEXT_FILE")
    if not context_file:
        raise RuntimeError("IGA_SCHEDULER_CONTEXT_FILE is not set")
    with open(context_file) as f:
        raw = json.load(f)
    return {
        "raw": raw,
        "run_id": raw.get("runId"),
        "definition": raw.get("definition", {}),
        "instance": raw.get("instance", {}),
        "scheduled_fire_time": raw.get("scheduledFireTime"),
        "attempt": raw.get("attempt", 1),
        "params": raw.get("params", {}),
        "param": ParameterReader(raw.get("params")),
        "iga_client": resolve_iga_client(),
    }
